ionoscloud/resource_apigateway.go

```diff
@@ -13,60 +18,150 @@ func resourceApiGateway() *schema.Resource {
+	}
+	gatewayID := *response.Id
+	d.SetId(gatewayID)
+	err = utils.WaitForResourceToBeReady(ctx, d, client.IsGatewayReady)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error checking status for API Gateway with ID: %v, error: %w", gatewayID, err))
+	}
+	if err := client.SetApiGatewayData(d, response); err != nil {
+		return diag.FromErr(err)
+	}
 	return nil
 }
 
-func resourceApiGatewayRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+func resourceApiGatewayRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(services.SdkBundle).ApiGatewayClient
+	gatewayID := d.Id()
+	gateway, resp, err := client.GetApiGatewayById(ctx, gatewayID)
+	if err != nil {
+		if resp.HttpNotFound() {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(fmt.Errorf("error reading API Gateway with ID: %v, error: %w", gatewayID, err))
+	}
+	log.Printf("[INFO] Successfully retrieved API Gateway with ID: %v, gateway info: %+v", gatewayID, gateway)
+	if err := client.SetApiGatewayData(d, gateway); err != nil {
+		return diag.FromErr(err)
+	}
 	return nil
 }
 
-func resourceApiGatewayUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+func resourceApiGatewayUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(services.SdkBundle).ApiGatewayClient
+	gatewayID := d.Id()
+
+	gateway, err := apigateway.GetGatewayDataEnsure(d)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error updating API Gateway: %w", err))
+	}
+
+	response, _, err := client.UpdateApiGateway(ctx, gatewayID, *gateway)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error updating API Gateway with ID: %v, error: %w", gatewayID, err))
+	}
+	err = utils.WaitForResourceToBeReady(ctx, d, client.IsGatewayReady)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error checking status for API Gateway with ID: %v after update, error: %w", gatewayID, err))
+	}
+	if err := client.SetApiGatewayData(d, response); err != nil {
+		return diag.FromErr(err)
+	}
 	return nil
 }
 
-func resourceApiGatewayDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+func resourceApiGatewayDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(services.SdkBundle).ApiGatewayClient
+	gatewayID := d.Id()
+	_, err := client.DeleteApiGateway(ctx, gatewayID)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error deleting API Gateway with ID: %v, error: %w", gatewayID, err))
+	}
+	err = utils.WaitForResourceToBeDeleted(ctx, d, client.IsGatewayDeleted)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("deletion check failed for API Gateway with ID: %v, error: %w", gatewayID, err))
+	}
 	return nil
 }
+
+func resourceApiGatewayImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
```

Comment:
This function is basically the same as `resourceApiGatewayRead`, maybe just call that one to avoid duplicated code.

Reply:
I'll make `Read` call `import` (the other way around of what you're suggesting) because it's easier to wrap an `error` into a `diags` than extracting an error from a diagnostics slice.